Source/Core/Common/Image.cpp

```diff
@@ -0,0 +1,183 @@
+// Copyright 2016 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <Common/Image.h>
+
+#include <csetjmp>
+#include <cstdlib>
+#include <cstring>
+#include <memory>
+#include <png.h>
+#include <vector>
+
+#include <Common/CommonTypes.h>
+#include <Common/Logging/Log.h>
+
+namespace Common
+{
+static inline int MultiplyAlpha(int alpha, int color)
+{
+  const int temp = (alpha * color) + 0x80;
+  return ((temp + (temp >> 8)) >> 8);
+}
+
+static void PremultiplyData(png_structp png, png_row_infop row_info, png_bytep data)
+{
+  for (png_size_t i = 0; i < row_info->rowbytes; i += 4, data += 4)
+  {
+    const png_byte alpha = data[3];
+    u32 w;
+
+    if (alpha == 0)
+    {
+      w = 0;
+    }
+    else
+    {
+      png_byte red = data[0];
+      png_byte green = data[1];
+      png_byte blue = data[2];
+
+      if (alpha != 0xff)
+      {
+        red = MultiplyAlpha(alpha, red);
+        green = MultiplyAlpha(alpha, green);
+        blue = MultiplyAlpha(alpha, blue);
+      }
+      w = (alpha << 24) | (red << 16) | (green << 8) | (blue << 0);
+    }
+
+    std::memcpy(data, &w, sizeof(w));
+  }
+}
+
+struct ReadProgress
+{
+  const u8* current;
+  const u8* end;
+};
+
+static void ReadCallback(png_structp png, png_bytep data, png_size_t size)
+{
+  ReadProgress* progress = reinterpret_cast<ReadProgress*>(png_get_io_ptr(png));
+  for (size_t i = 0; i < size; ++i)
+  {
+    if (progress->current >= progress->end)
+      png_error(png, "Read beyond end of file");
+
+    data[i] = *(progress->current);
+    ++(progress->current);
+  }
+}
+
+static void PNGErrorCallback(png_structp png, png_const_charp error_msg)
+{
+  ERROR_LOG(COMMON, "PNG loading error: %s", error_msg);
+  longjmp(png_jmpbuf(png), 1);
+}
+
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4611)
+// VS shows the following warning even though no C++ objects are destroyed in LoadPNG:
+// "warning C4611: interaction between '_setjmp' and C++ object destruction is non-portable"
+#endif
+
+bool LoadPNG(const std::vector<u8>& input, std::vector<u8>* data_out, u32* width_out,
```

Comment:
Perhaps it would likely be better API-wise to return a `std::optional<std::tuple<...>>` which allows both destructuring and success checking, which should be doable, as far as I can tell.

If not, then just disregard this.

Reply:
It should be doable, but it would make the calling code that we have right now more complicated. Is it worth it anyway?